### PR TITLE
Implemented Lectern

### PR DESCRIPTION
**Pull request by ipad54**
## Introduction
Implemented Lectern

### Relevant issues
Original PR is https://github.com/pmmp/PocketMine-MP/pull/4467, but it was closed

### Behavioural changes
N/A

## Tests
https://www.youtube.com/watch?v=lbory7FWuZw

### CODE REVIEW

**Comment by dktapps** [comment]
This also needs to be added to `VanillaBlocks` and `StringToItemParser`.

File: src/block/Lectern.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	/** @return $this */
+	public function setViewedPage(int $viewedPage) : self{
+		$this->viewedPage = $viewedPage;
+		return $this;
+	}
+
+	public function getBook() : ?WritableBookBase{
+		return $this->book instanceof WritableBookBase ? clone $this->book : $this->book;
```
Comment (dktapps) [inline src/block/Lectern.php]:
```suggestion
		return $this->book !== null ? clone $this->book : null;
```
for clearer reading

File: src/block/Lectern.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	/** @return $this */
+	public function setBook(?WritableBookBase $book) : self{
+		if($book instanceof WritableBookBase){
```
Comment (dktapps) [inline src/block/Lectern.php]:
```suggestion
		if($book !== null){
```

File: src/block/tile/Lectern.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	public const TAG_HAS_BOOK = "hasBook";
+	public const TAG_PAGE = "page";
+	public const TAG_TOTAL_PAGES = "totalPages";
+	public const TAG_BOOK = "book";
+
+	private int $viewedPage = 0;
+	private Item $book;
+
+	public function __construct(World $world, Vector3 $pos){
+		$this->book = VanillaItems::AIR();
```
Comment (dktapps) [inline src/block/tile/Lectern.php]: You can use `null` to represent no book, same as the block.

File: src/block/tile/Lectern.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	public function setViewedPage(int $viewedPage) : void{
+		$this->viewedPage = $viewedPage;
+	}
+
+	public function getBook() : ?WritableBookBase{
+		return $this->book instanceof WritableBookBase && !$this->book->isNull() ? $this->book : null;
+	}
+
+	public function setBook(?WritableBookBase $book) : void{
+		if(!$book instanceof WritableBookBase or $book->isNull()){
```
Comment (dktapps) [inline src/block/tile/Lectern.php]:
```suggestion
		if($book === null || $book->isNull()){
```

File: src/network/mcpe/handler/InGamePacketHandler.php [hunk 5]
```diff
@@ -856,6 +860,29 @@ public function handleLabTable(LabTablePacket $packet) : bool{
...
 	}
 
+	public function handleLecternUpdate(LecternUpdatePacket $packet) : bool{
+		//Drop book is handled on server side with an interact event.
+		//Total pages is gathered from book.
+		$pos = $packet->blockPosition;
+		$chunkX = $pos->getX() >> Chunk::COORD_BIT_SIZE;
+		$chunkZ = $pos->getZ() >> Chunk::COORD_BIT_SIZE;
+		$world = $this->player->getWorld();
+		if(!$world->isChunkLoaded($chunkX, $chunkZ) || !$world->isChunkGenerated($chunkX, $chunkZ) || $world->isChunkLocked($chunkX, $chunkZ)){
```
Comment (dktapps) [inline src/network/mcpe/handler/InGamePacketHandler.php]:
```suggestion
		if(!$world->isChunkLoaded($chunkX, $chunkZ) || $world->isChunkLocked($chunkX, $chunkZ)){
```
`isChunkGenerated()` is unnecessary; chunks which are not generated are also not loaded.

File: src/network/mcpe/handler/InGamePacketHandler.php [hunk 6]
```diff
@@ -856,6 +860,29 @@ public function handleLabTable(LabTablePacket $packet) : bool{
...
+			return false;
+		}
+		$lectern = $world->getBlockAt($pos->getX(), $pos->getY(), $pos->getZ());
+		if($lectern instanceof Lectern){
+			if($this->player->getPosition()->distance($lectern->getPosition()) <= 15){
```
Comment (dktapps) [inline src/network/mcpe/handler/InGamePacketHandler.php]: This check can be combined with the one above.

In addition, this should probably use `Player->canInteract()`.

File: src/network/mcpe/handler/InGamePacketHandler.php [hunk 7]
```diff
@@ -856,6 +860,29 @@ public function handleLabTable(LabTablePacket $packet) : bool{
...
+				$book = $lectern->getBook();
+				if($book instanceof WritableBookBase && count($book->getPages()) == $packet->totalPages && $packet->page >= 0 && $packet->page <= $packet->totalPages){
```
Comment (dktapps) [inline src/network/mcpe/handler/InGamePacketHandler.php]:
```suggestion
				if($book !== null && count($book->getPages()) === $packet->totalPages && $packet->page >= 0 && $packet->page < $packet->totalPages){
```

Off-by-one bug here. The page index should not be equal to the page count (the pages are zero-indexed).

File: src/block/tile/TileFactory.php [hunk 8]
```diff
@@ -73,6 +73,7 @@ public function __construct(){
 		$this->register(Sign::class, ["Sign", "minecraft:sign"]);
 		$this->register(Smoker::class, ["Smoker", "minecraft:smoker"]);
 		$this->register(Skull::class, ["Skull", "minecraft:skull"]);
+		$this->register(Lectern::class, ["Lectern", "minecraft:lectern"]);
```
Comment (dktapps) [inline src/block/tile/TileFactory.php]: Alphabetical order.

File: src/block/BlockFactory.php [hunk 9]
```diff
@@ -255,6 +256,7 @@ public function __construct(){
 		$this->registerAllMeta(new Lava(new BIDFlattened(Ids::FLOWING_LAVA, [Ids::STILL_LAVA], 0), "Lava", BlockBreakInfo::indestructible(500.0)));
 		$this->registerAllMeta(new Lever(new BID(Ids::LEVER, 0), "Lever", new BlockBreakInfo(0.5)));
 		$this->registerAllMeta(new Loom(new BID(Ids::LOOM, 0), "Loom", new BlockBreakInfo(2.5, BlockToolType::AXE)));
+		$this->registerAllMeta(new Lectern(new BID(Ids::LECTERN, 0, ItemIds::LECTERN, TileLectern::class), "Lectern", new BlockBreakInfo(2.5, BlockToolType::AXE)));
```
Comment (dktapps) [inline src/block/BlockFactory.php]: Alphabetical order (this should be above `Lever`).

File: src/block/tile/Lectern.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	public const TAG_BOOK = "book";
+
+	private int $viewedPage = 0;
+	private ?WritableBookBase $book = null;
+
+	public function readSaveData(CompoundTag $nbt) : void{
+		$this->viewedPage = $nbt->getInt(self::TAG_PAGE, 0);
+		if(($itemTag = $nbt->getCompoundTag(self::TAG_BOOK)) !== null){
+			$book = Item::nbtDeserialize($itemTag);
+			if($book instanceof WritableBookBase){
```
Comment (JavierLeon9966) [inline src/block/tile/Lectern.php]:
```suggestion
			if($book instanceof WritableBookBase && !$book->isNull()){
```

File: src/block/tile/Lectern.php [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+				$this->book = $book;
+			}
+		}
+	}
+
+	protected function writeSaveData(CompoundTag $nbt) : void{
+		$nbt->setByte(self::TAG_HAS_BOOK, $this->book !== null && !$this->book->isNull() ? 1 : 0);
```
Comment (JavierLeon9966) [inline src/block/tile/Lectern.php]:
```suggestion
		$nbt->setByte(self::TAG_HAS_BOOK, $this->book !== null ? 1 : 0);
```

File: src/block/tile/Lectern.php [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	public function getViewedPage() : int{
+		return $this->viewedPage;
+	}
+
+	public function setViewedPage(int $viewedPage) : void{
+		$this->viewedPage = $viewedPage;
+	}
+
+	public function getBook() : ?WritableBookBase{
+		return $this->book !== null && !$this->book->isNull() ? $this->book : null;
```
Comment (JavierLeon9966) [inline src/block/tile/Lectern.php]:
```suggestion
		return $this->book !== null ? clone $this->book : null;
```

File: src/block/tile/Lectern.php [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	public function setBook(?WritableBookBase $book) : void{
+		$this->book = $book;
```
Comment (JavierLeon9966) [inline src/block/tile/Lectern.php]:
```suggestion
		$this->book = $book !== null && !$book->isNull() ? clone $book : null;
```

File: src/block/tile/Lectern.php [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	protected function addAdditionalSpawnData(CompoundTag $nbt) : void{
+		$nbt->setByte(self::TAG_HAS_BOOK, $this->book !== null && !$this->book->isNull() ? 1 : 0);
```
Comment (JavierLeon9966) [inline src/block/tile/Lectern.php]:
```suggestion
		$nbt->setByte(self::TAG_HAS_BOOK, $this->book !== null ? 1 : 0);
```

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

File: src/block/Lectern.php [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+		return $this->book !== null ? clone $this->book : null;
+	}
+
+	/** @return $this */
+	public function setBook(?WritableBookBase $book) : self{
+		if($book !== null){
+			$book = clone $book;
+			$book->setCount(1);
+		}
+		$this->book = $book;
```
Comment (JavierLeon9966) [inline src/block/Lectern.php]:
```suggestion
		$this->book = $book !== null && !$book->isNull() ? (clone $book)->setCount(1) : null;
```
And you can remove the `if` statement from above

File: src/block/Lectern.php [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+		if($this->book === null && $item instanceof WritableBookBase){
+			$this->position->getWorld()->setBlock($this->position, $this->setBook($item));
+			$item->pop();
+			return true;
+		}
+		return false;
+	}
+
+	public function onAttack(Item $item, int $face, ?Player $player = null) : bool{
+		if($this->book !== null && !$this->book->isNull()){
```
Comment (JavierLeon9966) [inline src/block/Lectern.php]:
```suggestion
		if($this->book !== null){
```

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

**Comment by dktapps** [comment]
The collision box height in Bedrock is 0.9 blocks tall. The PR hasn't accounted for this.

**Comment by dktapps** [comment]
The hardness is also incorrect: it's 2.0 in Bedrock.

**Comment by dktapps** [comment]
The book must be included in the drops (for example, in case of explosions).

**Comment by dktapps** [comment]
A block will be placed if right-clicking on a lectern while holding a block. This isn't supposed to happen.

**Pull request merged**